pag-parser2/src/nf/semact.rs

```diff
@@ -8,42 +8,111 @@
 
 use std::collections::HashMap;
 
-use quote::format_ident;
-use syn::{parse_quote, Expr, ExprCall, Stmt, Type};
-
 use super::Tag;
+use syn::{parse_quote, Expr, Type};
 
 pub type SemActTable = HashMap<Tag, SemAct>;
-pub struct SemAct {
-    /// Identifier of the semantic action routine.
-    function: Expr,
-    /// Type annotation
-    ty: Option<Type>,
-    /// Number of arguments
-    arity: usize,
+
+///
+/// ```
+/// trait Collector<T> {
+///     pub type Output;
+///     fn finalize(self) -> Self::Output;
+///     fn collect(&mut self, data: T);
+/// }
```

Comment:
The trait in my mind is like:

```rust
trait Collector<T>: Default {
    fn collect(&mut self, data: T);
}
```

And for `x+` we do `default` + `collect` instead of `from` + `collect` because std containers don't have `From<T>`.

Reply:
Sure. I think that will be good enough and it also simplifies the design.